### PR TITLE
Verify registry signature

### LINKED ISSUE #1266: Implement allowlist signature verification
Snap-registry allow list will be signed using private keys managed by the npm-publishers team.

## Outcome
* The extension downloads the allow list json file and a second file which contains the signature of the aforementioned json file.
* The extension verifies the allow list signature.
* The extension displays a user error if the signature is invalid.

### PR DESCRIPTION
**Pull request by Mrtenz**
Closes #1266.

This sets up `snaps-registry` signature verification, using a configurable public key.

### CODE REVIEW

File: packages/snaps-controllers/src/snaps/registry/json.ts [hunk 1]
```diff
@@ -12,17 +20,29 @@ import {
...
-const SNAP_REGISTRY_URL =
-  'https://cdn.jsdelivr.net/gh/MetaMask/snaps-registry@main/src/registry.json';
+export const SNAP_REGISTRY_URL =
+  'https://cdn.jsdelivr.net/gh/MetaMask/snaps-registry@gh-pages/latest/registry.json';
+
+export const SNAP_REGISTRY_SIGNATURE_URL =
+  'https://cdn.jsdelivr.net/gh/MetaMask/snaps-registry@gh-pages/latest/signature.json';
+
+// TODO: Set public key.
+const SNAP_REGISTRY_PUBLIC_KEY =
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/snaps/registry/json.ts:30]: Should be configurable via constructor imo

File: packages/snaps-controllers/src/snaps/registry/json.ts [hunk 2]
```diff
@@ -115,4 +141,42 @@ export class JsonSnapsRegistry implements SnapsRegistry {
...
+
+  /**
+   * Verify the signature of the registry.
+   *
+   * @param database - The registry database.
+   * @param signature - The signature of the registry.
+   * @throws If the signature is invalid.
+   * @private
+   */
+  async #verifySignature(database: string, signature: string) {
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/snaps/registry/json.ts:153]:
```suggestion
  async #verifySignature(database: SnapsRegistryDatabase, signature: string) {
```

File: packages/snaps-controllers/src/snaps/registry/json.test.ts [hunk 3]
```diff
@@ -37,11 +38,23 @@ const MOCK_DATABASE: SnapsRegistryDatabase = {
   ],
 };
 
+const MOCK_SIGNATURE =
+  '0x3045022100cc049732d4cc8b888162b8b998c9beefcc1de8c4489594c7c504d4aa031223af02206835834ba2b7bec45b2a9e1e72312d69445d6a0f5590a2408bb13931a14f6c8a';
+const MOCK_PUBLIC_KEY =
+  '0x034ca27b046507d1a9997bddc991b56d96b93d4adac3a96dfe01ce450bfb661455';
+const MOCK_SIGNATURE_FILE = {
+  signature: MOCK_SIGNATURE,
+  publicKey: MOCK_PUBLIC_KEY,
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/snaps/registry/json.test.ts:47]: Why do we need the `MOCK_PUBLIC_KEY` here? 🤔

File: packages/snaps-controllers/src/snaps/registry/json.ts (hunk 1 above)
Comment (FrederikBolding) [reply packages/snaps-controllers/src/snaps/registry/json.ts:30]: That way we can disable it as well by not passing in a pub key

File: packages/snaps-controllers/src/snaps/registry/json.test.ts (hunk 3 above)
Comment (Mrtenz) [reply packages/snaps-controllers/src/snaps/registry/json.test.ts:46]: That's the format used in MetaMask/snaps-registry#18. It's not used, but seems better to mock with "real" data.

File: packages/snaps-controllers/src/snaps/registry/json.ts [hunk 4]
```diff
@@ -72,7 +83,11 @@ export class JsonSnapsRegistry extends BaseController<
   SnapsRegistryState,
   SnapsRegistryMessenger
 > {
-  #url: string;
+  #url: JsonSnapsRegistryUrl;
+
+  #publicKey?: Hex;
+
+  #database: SnapsRegistryDatabase | null = null;
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/snaps/registry/json.ts:90]:
```suggestion
```

File: packages/snaps-controllers/src/snaps/registry/json.ts [hunk 5]
```diff
@@ -127,21 +147,24 @@ export class JsonSnapsRegistry extends BaseController<
...
       }
-      const db = await response.json();
+
       this.update((state) => {
-        state.database = db;
+        state.database = JSON.parse(database);
         state.lastUpdated = Date.now();
       });
-    } catch {
+    } catch (error) {
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/snaps/registry/json.ts:167]:
```suggestion
    } catch {
```

File: packages/snaps-controllers/src/snaps/registry/json.ts [hunk 6]
```diff
@@ -127,21 +147,24 @@ export class JsonSnapsRegistry extends BaseController<
...
     }
+
     try {
-      const response = await this.#fetchFunction(this.#url);
-      if (!response.ok) {
-        throw new Error('Failed to fetch Snaps registry.');
+      const database = await this.#safeFetch(this.#url.registry);
+      const signature = await this.#safeFetch(this.#url.signature);
+
+      if (this.#publicKey) {
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/snaps/registry/json.ts:159]: Can we skip fetching the signature if we don't have a public key?

**Review by FrederikBolding — APPROVED** [review]

**Review by FrederikBolding — APPROVED** [review]

**Pull request merged**